File: assign5-java/FileHandle.java
//****************************************************************************
// File  : FileHandle.java
// Author: Keith T. Fancher
// Date  : 03/17/04
// Class : CS3490
//
// Purpose: The FileHandle class used by the assembler.  It handles files.
//****************************************************************************


import java.io.*;


class FileHandle
{
  private String inputFile;
  private String outputFile;

  private boolean badFileName()
  {
    // make sure it ends with .asm 'n such
    if(!inputFile.endsWith(".asm") || inputFile.length() < 5)
      return true;
    return false;
  }

  private void buildOutputFileName()
  {
    // this is a nifty statement
    outputFile = new String((inputFile.substring(0, inputFile.lastIndexOf('.') + 1)).concat("hex"));
  }


  public FileHandle(String filename)
  {
    inputFile = new String(filename);
  }

  public FileReader openInputFile() throws IOException
  {
    if(badFileName())
    {
      System.err.println("Filename must be at least 5 characters long and end with .asm");
      System.err.println("Usage: Assemble <filename>");
      System.exit(1);
    }

    // If you put this in a try block and catch the exception you won't have anything to
    // return and Java gets mad, even though it NEVER GETS TO THAT PART OF THE CODE...
    // It's especially annoying because there doesn't seem to be any way to specify an empty
    // FileReader, which means manual error checking is pretty much out of the question.
    // Why oh why can't it just trust me to handle my own friggin' exceptions???
    return (new FileReader(inputFile));
  }

  public FileWriter openOutputFile() throws IOException
  {
    buildOutputFileName();
    return (new FileWriter(outputFile));
  }

  public void closeFiles(FileReader infile, FileWriter outfile) throws IOException
  {
    infile.close();
    outfile.close();
  }

  public String getOutputFileName()
  {
    return outputFile;
  }
}

File: assign2-cpp/FileHandle.cpp
using namespace std;
#include <iostream>
#include <fstream>
#include "FileHandle.h"

FileHandle::FileHandle(char * filename)
{
    inputFile = filename;
}
              
void FileHandle::openInputFile(ifstream & inf)
{
    if (badFileName())
    {
        cerr << inputFile << " must be a least 5 characters and end with "
             << "a .asm extension\n";
        cerr << "Usage: assemble <filename>\n";
        exit(1);
    }
    inf.open(inputFile.c_str());
    if (inf.fail())
    {
        cerr << "Can not open: " << inputFile << endl;
        cerr << "Usage: ma <filename>\n";
        exit(1);
    }
}

bool FileHandle::badFileName()
{
    int i;
    i = inputFile.find(".asm", 1);
    //make sure that the .asm is at the end of the string
    if (inputFile.size() < 5 || i != inputFile.size() - 4)
        return true;
    else
        return false;
}

void FileHandle::buildOutputFileName()
{
    int i;
    //find the location of the .asm
    i = inputFile.find(".asm", 1);
    outputFile = inputFile.replace(i, 4, ".hex");
}


void FileHandle::openOutputFile(ofstream & outf)
{
    buildOutputFileName();
    outf.open(outputFile.c_str());
    if (outf.fail())
    {
        cerr << "Can not open output file: " << outputFile << endl;
        cerr << "Usage: ma <filename>\n";
        exit(1);
    }
}

void FileHandle::closeFiles(ifstream & inf, ofstream & outf)
{
   inf.close();
   outf.close();
}
